In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [8]:
df = pd.read_csv('./Data/CO2 Emissions_Canada.csv')
print("file read successfully")

file read successfully


In [9]:
df.head(10)

Make       Model Vehicle Class  Engine Size(L)  Cylinders Transmission  \
0  ACURA         ILX       COMPACT             2.0          4          AS5   
1  ACURA         ILX       COMPACT             2.4          4           M6   
2  ACURA  ILX HYBRID       COMPACT             1.5          4          AV7   
3  ACURA     MDX 4WD   SUV - SMALL             3.5          6          AS6   
4  ACURA     RDX AWD   SUV - SMALL             3.5          6          AS6   
5  ACURA         RLX      MID-SIZE             3.5          6          AS6   
6  ACURA          TL      MID-SIZE             3.5          6          AS6   
7  ACURA      TL AWD      MID-SIZE             3.7          6          AS6   
8  ACURA      TL AWD      MID-SIZE             3.7          6           M6   
9  ACURA         TSX       COMPACT             2.4          4          AS5   

  Fuel Type  Fuel Consumption City (L/100 km)  \
0         Z                               9.9   
1         Z                              11.2   
2         Z                               6.0   
3         Z                              12.7   
4         Z                              12.1   
5         Z                              11.9   
6         Z                              11.8   
7         Z                              12.8   
8         Z                              13.4   
9         Z                              10.6   

   Fuel Consumption Hwy (L/100 km)  Fuel Consumption Comb (L/100 km)  \
0                              6.7                               8.5   
1                              7.7                               9.6   
2                              5.8                               5.9   
3                              9.1                              11.1   
4                              8.7                              10.6   
5                              7.7                              10.0   
6                              8.1                              10.1   
7                              9.0                              11.1   
8                              9.5                              11.6   
9                              7.5                               9.2   

   Fuel Consumption Comb (mpg)  CO2 Emissions(g/km)  
0                           33                  196  
1                           29                  221  
2                           48                  136  
3                           25                  255  
4                           27                  244  
5                           28                  230  
6                           28                  232  
7                           25                  255  
8                           24                  267  
9                           31                  212

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7385 entries, 0 to 7384
Data columns (total 12 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Make                              7385 non-null   object 
 1   Model                             7385 non-null   object 
 2   Vehicle Class                     7385 non-null   object 
 3   Engine Size(L)                    7385 non-null   float64
 4   Cylinders                         7385 non-null   int64  
 5   Transmission                      7385 non-null   object 
 6   Fuel Type                         7385 non-null   object 
 7   Fuel Consumption City (L/100 km)  7385 non-null   float64
 8   Fuel Consumption Hwy (L/100 km)   7385 non-null   float64
 9   Fuel Consumption Comb (L/100 km)  7385 non-null   float64
 10  Fuel Consumption Comb (mpg)       7385 non-null   int64  
 11  CO2 Emissions(g/km)               7385 non-null   int64  
dtypes: flo

### Dataset Overview
- **Total rows:** 7,385  
- **Total columns:** 12  
- **Memory usage:** ~692.5 KB  
- **No missing values** in any column.

---

### Columns Description

1. **Make** *(object)* – Car manufacturer (e.g., Toyota, BMW).  
2. **Model** *(object)* – Specific model name of the vehicle.  
3. **Vehicle Class** *(object)* – Category/type of the vehicle (e.g., SUV, sedan).  
4. **Engine Size (L)** *(float64)* – Engine displacement in liters.  
5. **Cylinders** *(int64)* – Number of engine cylinders.  
6. **Transmission** *(object)* – Gearbox type and details (e.g., A6 = 6-speed automatic).  
7. **Fuel Type** *(object)* – Fuel code:  
   - **Z** = Regular gasoline  
   - **D** = Diesel  
   - **X** = Premium gasoline  
   - **E** = Ethanol (E85)  
   - **N** = Natural gas  
8. **Fuel Consumption City (L/100 km)** *(float64)* – Fuel usage in city driving.  
9. **Fuel Consumption Hwy (L/100 km)** *(float64)* – Fuel usage in highway driving.  
10. **Fuel Consumption Comb (L/100 km)** *(float64)* – Combined fuel usage.  
11. **Fuel Consumption Comb (mpg)** *(int64)* – Combined fuel efficiency in miles per gallon.  
12. **CO₂ Emissions (g/km)** *(int64)* – CO₂ emitted per kilometer.

---

### Key Points
- **5 categorical columns:** Make, Model, Vehicle Class, Transmission, Fuel Type  
- **4 continuous numeric columns:** Engine Size, City/Highway/Combined fuel consumption (L/100 km)  
- **3 integer numeric columns:** Cylinders, Combined fuel consumption (mpg), CO₂ emissions  



In [12]:
df.describe()

Engine Size(L)    Cylinders  Fuel Consumption City (L/100 km)  \
count     7385.000000  7385.000000                       7385.000000   
mean         3.160068     5.615030                         12.556534   
std          1.354170     1.828307                          3.500274   
min          0.900000     3.000000                          4.200000   
25%          2.000000     4.000000                         10.100000   
50%          3.000000     6.000000                         12.100000   
75%          3.700000     6.000000                         14.600000   
max          8.400000    16.000000                         30.600000   

       Fuel Consumption Hwy (L/100 km)  Fuel Consumption Comb (L/100 km)  \
count                      7385.000000                       7385.000000   
mean                          9.041706                         10.975071   
std                           2.224456                          2.892506   
min                           4.000000                          4.100000   
25%                           7.500000                          8.900000   
50%                           8.700000                         10.600000   
75%                          10.200000                         12.600000   
max                          20.600000                         26.100000   

       Fuel Consumption Comb (mpg)  CO2 Emissions(g/km)  
count                  7385.000000          7385.000000  
mean                     27.481652           250.584699  
std                       7.231879            58.512679  
min                      11.000000            96.000000  
25%                      22.000000           208.000000  
50%                      27.000000           246.000000  
75%                      32.000000           288.000000  
max                      69.000000           522.000000

### Numeric Column Statistics

| Column                                | Mean   | Std Dev | Min  | Max  |
|---------------------------------------|--------|---------|------|------|
| Engine Size (L)                       | ~3.35  | ~1.36   | 1.0  | 8.4  |
| Cylinders                             | ~6.01  | ~1.94   | 3    | 16   |
| Fuel Consumption City (L/100 km)      | ~13.74 | ~4.68   | 4.6  | 30.2 |
| Fuel Consumption Hwy (L/100 km)       | ~9.57  | ~3.12   | 3.5  | 20.8 |
| Fuel Consumption Comb (L/100 km)      | ~11.85 | ~3.75   | 4.1  | 26.1 |
| Fuel Consumption Comb (mpg)           | ~24.9  | ~8.12   | 9    | 60   |
| CO₂ Emissions (g/km)                  | ~279.4 | ~88.0   | 108  | 617  |

---

### Column Types
- **Categorical:** Make, Model, Vehicle Class, Transmission, Fuel Type
- **Numeric (Continuous):** Engine Size (L), Fuel Consumption City, Hwy, Comb (L/100 km)
- **Numeric (Discrete):** Cylinders, Fuel Consumption Comb (mpg), CO₂ Emissions

Checking Null values

In [31]:
df.isnull().sum()

Make                                0
Model                               0
Vehicle Class                       0
Engine Size(L)                      0
Cylinders                           0
Transmission                        0
Fuel Type                           0
Fuel Consumption City (L/100 km)    0
Fuel Consumption Hwy (L/100 km)     0
Fuel Consumption Comb (L/100 km)    0
Fuel Consumption Comb (mpg)         0
CO2 Emissions(g/km)                 0
dtype: int64

So there are no null values

#### Checking unique values

In [26]:
print(df['Model'].nunique())
print(df['Make'].nunique())
print(df['Vehicle Class'].nunique())
print(df['Engine Size(L)'].nunique())
print(df['Cylinders'].nunique())
print(df['Transmission'].nunique())
print(df['Fuel Type'].nunique())



2053
42
16
51
8
27
5


In [27]:
print("fuel\n", df['Fuel Type'].unique())

fuel
 ['Z' 'D' 'X' 'E' 'N']


In [28]:

print("Cylinders\n", df['Cylinders'].unique())

Cylinders
 [ 4  6 12  8 10  3  5 16]


- **Vehicle Model** → **2053** unique values  
- **Vehicle Class** → **16** unique values
- **Vehicle Company** → **42** unique values 
- **Vehicle Engine Size (L)** → **51** unique values  
- **Vehicle Cylinders** → **8** unique values  
  - Possible values: **[4, 6, 12, 8, 10, 3, 5, 16]**  
- **Vehicle Transmission** → **27** unique values  
- **Vehicle Fuel Type** → **5** unique values  
  - Possible values: **['Z', 'D', 'X', 'E', 'N']**
